File: web/packages/dtx-core/tests/zz.test.ts
import { describe, it, expect } from 'vitest';
import { decodeZz } from '../src/parser/zz.js';

describe('decodeZz', () => {
  it('decodes 00 as 0', () => {
    expect(decodeZz('00')).toBe(0);
  });
  it('decodes 01 as 1', () => {
    expect(decodeZz('01')).toBe(1);
  });
  it('decodes 0A as 10', () => {
    expect(decodeZz('0A')).toBe(10);
  });
  it('decodes 10 as 36', () => {
    expect(decodeZz('10')).toBe(36);
  });
  it('decodes ZZ as 36*36 - 1 = 1295', () => {
    expect(decodeZz('ZZ')).toBe(35 * 36 + 35);
  });
  it('tolerates lowercase', () => {
    expect(decodeZz('ab')).toBe(decodeZz('AB'));
  });
  it('throws on non-base36', () => {
    expect(() => decodeZz('!!')).toThrow();
  });
  it('throws on wrong length', () => {
    expect(() => decodeZz('0')).toThrow();
    expect(() => decodeZz('000')).toThrow();
  });
});

File: native/quest-spatial/app/src/test/kotlin/com/dtxmania/quest/dtxcore/parser/ZzTest.kt
package com.dtxmania.quest.dtxcore.parser

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.Test

/**
 * Ported 1:1 from `web/packages/dtx-core/tests/zz.test.ts`. Each `it(...)`
 * block in the Vitest source maps to one [Test] here with the same
 * fixture values.
 */
class ZzTest {
    @Test fun `decodes 00 as 0`() {
        assertEquals(0, decodeZz("00"))
    }

    @Test fun `decodes 01 as 1`() {
        assertEquals(1, decodeZz("01"))
    }

    @Test fun `decodes 0A as 10`() {
        assertEquals(10, decodeZz("0A"))
    }

    @Test fun `decodes 10 as 36`() {
        assertEquals(36, decodeZz("10"))
    }

    @Test fun `decodes ZZ as 36 squared minus 1`() {
        assertEquals(35 * 36 + 35, decodeZz("ZZ"))
    }

    @Test fun `tolerates lowercase`() {
        assertEquals(decodeZz("AB"), decodeZz("ab"))
    }

    @Test fun `throws on non-base36`() {
        assertThrows<IllegalArgumentException> { decodeZz("!!") }
    }

    @Test fun `throws on wrong length`() {
        assertThrows<IllegalArgumentException> { decodeZz("0") }
        assertThrows<IllegalArgumentException> { decodeZz("000") }
    }
}
